fix(ivy): support injecting ChangeDetectorRef on templates

JoostK opened the pull request:
Previously, using a pipe in an input binding on an ng-template would
evaluate the pipe in the context of node that was processed before the
template. This caused the retrieval of e.g. ChangeDetectorRef to be
incorrect, resulting in one of the following bugs depending on the
template's structure:

1. If the template was at the root of a view, the previously processed
node would be the component's host node outside of the current view.
Accessing that node in the context of the current view results in a crash.
2. For templates not at the root, the ChangeDetectorRef injected into the
pipe would correspond with the previously processed node. If that node
hosts a component, the ChangeDetectorRef would not correspond with the
view that the ng-template is part of.

The solution to the above problem is two-fold:

1. Template compilation is adjusted such that the template instruction
is emitted before any instructions produced by input bindings, such as
pipes. This ensures that pipes are evaluated in the context of the
template's container node.
2. A ChangeDetectorRef can be requested for container nodes.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

JoostK commented on packages/core/test/render3/di_spec.ts:
```diff
@@ -1718,6 +1718,64 @@ describe('di', () => {
+            template: function(rf: RenderFlags, ctx: MyApp) {
+              if (rf & RenderFlags.Create) {
+                template(0, C1, 3, 1, null, [AttributeMarker.SelectOnly, 'ngIf']);
+                pipe(1, 'async');
```
**NOTE:** The above template is **NOT** what the compiler generates. Instead, the generated code for the given template would be:

```
if (rf & RenderFlags.Create) {
  pipe(1, 'async'); // swapped to be positioned before the `template` instruction
  template(0, C1, 3, 1, null, [AttributeMarker.SelectOnly, 'ngIf']);
}
```

This however causes the runtime to crash, as during the creation of the `AsyncPipe` instance the `ChangeDetectorRef` reads its component using `getComponentViewByIndex` but at that point, the view state is not as expected.

```
Failures:
1) di Special tokens ChangeDetectorRef should inject host component ChangeDetectorRef into directives on templates
  Message:
    TypeError: Cannot read property 'length' of null
  Stack:
        at <Jasmine>
        at Object.getComponentViewByIndex (packages/core/test/packages/core/src/render3/util.ts:111:20)
        at createViewRef (packages/core/test/packages/core/src/render3/view_engine_compatibility.ts:360:27)
        at Object.injectChangeDetectorRef [as SWITCH_CHANGE_DETECTOR_REF_FACTORY__POST_R3__] (packages/core/test/packages/core/src/render3/view_engine_compatibility.ts:345:10)
        at <Jasmine>
        at Object.getOrCreateInjectable (packages/core/test/packages/core/src/render3/di.ts:307:21)
        at Object.directiveInject (packages/core/test/packages/core/src/render3/instructions.ts:2829:10)
        at Object.factory (packages/core/test/packages/core/test/render3/di_spec.ts:1736:57)
        at Object.pipe (packages/core/test/packages/core/src/render3/pipe.ts:43:32)
        at template (packages/core/test/packages/core/test/render3/di_spec.ts:1767:17)
        at detectChangesInternal (packages/core/test/packages/core/src/render3/instructions.ts:2529:5)
        at componentRefresh (packages/core/test/packages/core/src/render3/instructions.ts:2175:5)
        at refreshChildComponents (packages/core/test/packages/core/src/render3/instructions.ts:155:7)
        at Object.refreshDescendantViews (packages/core/test/packages/core/src/render3/instructions.ts:94:3)
        at Object.renderComponent (packages/core/test/packages/core/src/render3/component.ts:136:5)
        at new ComponentFixture (packages/core/test/packages/core/test/render3/render_util.ts:159:22)
        at UserContext.<anonymous> (packages/core/test/packages/core/test/render3/di_spec.ts:1779:21)
```

@kara @mhevery I don't know if the compiler needs to be fixed to swap the instructions, or if the runtime needs to be updated somehow. Please advise :)

pkozlowski-opensource replied:
@JoostK so we are trying to _not_ to write runtime - only tests any more. Could you please remove this runtime test and convert it into a `TestBed` one? This way we will have a reproduce scenario for both a compiler and a runtime.

As soon as we've got a minimal `TestBed` reproduce scenario we will take your PR (alongside with #28587) and dig into it.

Thnx for looking into it an PRing!

JoostK replied:
@pkozlowski-opensource ah yes this test was created when using runtime instructions was still the norm. I will look into creating a `TestBed` soon.

JoostK committed:
fix(ivy): support injecting ChangeDetectorRef on templates

Previously, using a pipe in an input binding on an ng-template would
evaluate the pipe in the context of node that was processed before the
template. This caused the retrieval of e.g. ChangeDetectorRef to be
incorrect, resulting in one of the following bugs depending on the
template's structure:

1. If the template was at the root of a view, the previously processed
node would be the component's host node outside of the current view.
Accessing that node in the context of the current view results in a crash.
2. For templates not at the root, the ChangeDetectorRef injected into the
pipe would correspond with the previously processed node. If that node
hosts a component, the ChangeDetectorRef would not correspond with the
view that the ng-template is part of.

The solution to the above problem is two-fold:

1. Template compilation is adjusted such that the template instruction
is emitted before any instructions produced by input bindings, such as
pipes. This ensures that pipes are evaluated in the context of the
template's container node.
2. A ChangeDetectorRef can be requested for container nodes.

Fixes #28587

JoostK commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -0,0 +1,42 @@
+      TestBed.configureTestingModule({declarations: [MyApp, MyPipe], imports: [CommonModule]});
+      const fixture = TestBed.createComponent(MyApp);
+      fixture.detectChanges();
+      expect((pipeInstance !.cdr as ViewRef<MyApp>).context).toBe(fixture.componentInstance);
```
I am not too happy about referring to implementation details from within the acceptance `TestBed` suite, so this should probably be done differently. @pkozlowski-opensource WDYT?

mhevery commented:
http://test/OCL:233690116:BASE:233690154:1550038053377:835cae0a
